### ISSUE TITLE
certbot container fails with an error

### ISSUE THREAD

**Issue by mittkkoo** [evgeniy-khist/letsencrypt-docker-compose#56]
I tried to execute the script with the option Multiple Docker Compose projects, because on that server I have Keycloak and PostgreSQL running in other containers. 
```
? What's your domain name (e.g. example.com)? xxx.xx.xx
? What's your email for registration and recovery contact? xxx@xxx.xx
? Want to have 'www' subdomain (e.g. www.example.com)? No
? Want to obtain a test certificate from a staging server? Yes
? What is the RSA key size in bits? 4096
? How do you want to configure Nginx? Reverse proxy
? Does the upstream server run as a Docker container on the same host? Yes
? What is the address of the proxied server (e.g. example-backend:8080)? keycloak_demo:8080
? Enable WebSocket proxying? No
? Want to add another domain? No
? What is the DH parameters size in bits? 2048
? Use Gzip? No
? Are the entered data correct? Yes
```

However the certbot container fails to start.

```
dependency failed to start: container letsencrypt-docker-compose-certbot-1 is unhealthy
```

**Comment by evgeniy-khist** [comment]
Certbot container logs are required to try to find out what is the problem: `docker compose logs certbot`

**Comment by mittkkoo** [comment]
Sure!
```
letsencrypt-docker-compose-certbot-1  | Obtaining the certificate for domain key.xxx.xx
letsencrypt-docker-compose-certbot-1  | Testing on staging environment enabled
letsencrypt-docker-compose-certbot-1  | Using email xx@xxx.xx
letsencrypt-docker-compose-certbot-1  | RSA key size is 4096
letsencrypt-docker-compose-certbot-1  | Saving debug log to /var/log/letsencrypt/letsencrypt.log
letsencrypt-docker-compose-certbot-1  | Plugins selected: Authenticator webroot, Installer None
letsencrypt-docker-compose-certbot-1  | Account registered.
letsencrypt-docker-compose-certbot-1  | Requesting a certificate for key.xxx.xx
letsencrypt-docker-compose-certbot-1  | Performing the following challenges:
letsencrypt-docker-compose-certbot-1  | http-01 challenge for key.xxx.xx
letsencrypt-docker-compose-certbot-1  | Using the webroot path /var/www/certbot/key.xxx.xx for all unmatched domains.
letsencrypt-docker-compose-certbot-1  | Waiting for verification...
letsencrypt-docker-compose-certbot-1  | Challenge failed for domain key.xxx.xx
letsencrypt-docker-compose-certbot-1  | http-01 challenge for key.xxx.xx
letsencrypt-docker-compose-certbot-1  | 
letsencrypt-docker-compose-certbot-1  | Certbot failed to authenticate some domains (authenticator: webroot). The Certificate Authority reported these problems:
letsencrypt-docker-compose-certbot-1  |   Domain: key.xxx.xx
letsencrypt-docker-compose-certbot-1  |   Type:   connection
letsencrypt-docker-compose-certbot-1  |   Detail: 2x.xx.xx.xx: Fetching http://key.xxx.xx/.well-known/acme-challenge/FTlE3rx78GzcL0klOFXf7dmAnp-mxxx: Timeout during connect (likely firewall problem)
letsencrypt-docker-compose-certbot-1  | 
letsencrypt-docker-compose-certbot-1  | Hint: The Certificate Authority failed to download the temporary challenge files created by Certbot. Ensure that the listed domains serve their content from the provided --webroot-path/-w and that files created there can be downloaded from the internet.
letsencrypt-docker-compose-certbot-1  | 
letsencrypt-docker-compose-certbot-1  | Cleaning up challenges
letsencrypt-docker-compose-certbot-1  | Some challenges have failed.
letsencrypt-docker-compose-certbot-1  | Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /var/log/letsencrypt/letsencrypt.log or re-run Certbot with -v for more details.

```

**Comment by evgeniy-khist** [comment]
```
Certbot failed to authenticate some domains (authenticator: webroot). The Certificate Authority reported these problems:
Detail: 2x.xx.xx.xx: Fetching http://key.xxx.xx/.well-known/acme-challenge/FTlE3rx78GzcL0klOFXf7dmAnp-mxxx: Timeout during connect (likely firewall problem)
```

```bash
./cli.sh up --dry-run
```

or

```bash
```

**Comment by evgeniy-khist** [comment]
Certbot failed to authenticate domain. Here is the error message:

```
Certbot failed to authenticate some domains (authenticator: webroot). The Certificate Authority reported these problems:
Detail: 2x.xx.xx.xx: Fetching http://key.xxx.xx/.well-known/acme-challenge/FTlE3rx78GzcL0klOFXf7dmAnp-mxxx: Timeout during connect (likely firewall problem)
```

Double check your DNS configuration.
Make sure that the website is accessible with self-signed certificates. Run the Docker Compose project in "Dry Run" mode (without running Certbot):

```bash
./cli.sh up --dry-run
```

or

```bash
DRY_RUN=true docker compose up -d
```

and then try to open any page of you website.

Browser will complain that certificate is self-signed, accept it and proceed to make sure that the page can be accessed.

**Comment by mittkkoo** [comment]
The dry run passes without errors! I also can access the landing UI page of Keycloak (after accepting the exception), which is now https. However not the admin area. The browser logs say
```
Content-Security-Policy: The page's settings blocked the loading of a resource at http://key.xx.xx/realms/master/protocol/openid-connect/3p-cookies/step1.html
```

**Comment by evgeniy-khist** [comment]
Please verify, that the data for Let's Encrypt ACME challenge can be accessed.

Run the project in the dry run mode (without actually running Certbot):

```bash
./cli.sh up --dry-run
```

Specify your domain:

```bash
domain=your.domain.com
```

Create a file that will emulate ACME challenge:

```bash
docker compose exec certbot mkdir -p /var/www/certbot/${domain}/.well-known/acme-challenge/
docker compose exec certbot sh -c "echo $(date) > /var/www/certbot/${domain}/.well-known/acme-challenge/test-challenge.txt"
```

Verify that the test file was created:

```bash
docker compose exec certbot cat /var/www/certbot/${domain}/.well-known/acme-challenge/test-challenge.txt
```

Now, verify that the Nginx is serving this test file emulating ACME challenge and it can be accessed:

```bash
curl http://${domain}/.well-known/acme-challenge/test-challenge.txt
```

The output of the `curl` command must be equal to the content of the `test-challenge.txt` file.

**Comment by mittkkoo** [comment]
The last step failed
```
curl: (28) Failed to connect to key.xxx.xx port 80 after 131066 ms: Connection timed out
```
However if I enter the same URL in a browser it will correctly return the content of the txt, also through http. How can that be?
```
Mon Jul 17 09:53:09 AM UTC 2023
```

**Comment by evgeniy-khist** [comment]
Check your network settings to see if there is a proxy server configured.

**Comment by evgeniy-khist** [comment]
Can you access the `http://${domain}/.well-known/acme-challenge/test-challenge.txt` from the browser but from another device, for example smartphone?

**Comment by mittkkoo** [comment]
I discovered that I had location / country blocking activated in the firewall. When I switched that off the http-01 challenge passed without error. So I completed Step 4 of the installation guide and the certificate was issued by STAGING. However still `curl --insecure https://domain` fails with a timeout error. Despite of that I switched to production env. Now the strange thing is that the certificate is self-signed (Issuer name Common Name = my domain). And again, `curl domain` fails with a timeout.

Yes, I tried `http://${domain}/.well-known/acme-challenge/test-challenge.txt`from different devices and networks and I can access it through the browser.

**Comment by evgeniy-khist** [comment]
`docker-compose` is V1 and not supported anymore. Consider updating to V2 `docker compose`.
Did you succeed in getting production Let's Encrypt certificates?

**Comment by mittkkoo** [comment]
> `docker-compose` is V1 and not supported anymore. Consider updating to V2 `docker compose`. Did you succeed in getting production Let's Encrypt certificates?

The server runs on Ubuntu and I choosed to install from docker.io, because [this reasoning](https://stackoverflow.com/questions/45023363/what-is-docker-io-in-relation-to-docker-ce-and-docker-ee-now-called-mirantis-k/57678382#57678382) sounds convincing to me. And with that I have to use docker-compose.

Yes, I got a let's encrypt certificate. Then I also turned on a Cloudflare proxy and now the issuer of the certificate is sni.cloudflaressl.com. The certificate is verified by Cloudflare. I'm somewhat surprised that it works like that, but maybe I need to dive deeper to understand the process. Anyway, your hints were extremely valuable for finding the problem!
